domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt

```diff
@@ -0,0 +1,291 @@
+          AsyncResult.failed(
+            ExplorationCheckpointNotFoundException(
+              "No saved checkpoints in $CACHE_NAME for profileId ${profileId.internalId}."
+            )
+          )
+        } else {
+          val explorationCheckpointDetails = ExplorationCheckpointDetails.newBuilder()
+            .setExplorationId(oldestCheckpoint.key)
+            .setExplorationTitle(oldestCheckpoint.value.explorationTitle)
+            .setExplorationVersion(oldestCheckpoint.value.explorationVersion)
+            .build()
+          AsyncResult.success(explorationCheckpointDetails)
+        }
+      }
+  }
+
+  /** deletes the saved checkpoint for a specified explorationId and profileId. */
+  fun deleteSavedExplorationCheckpoint(
+    profileId: ProfileId,
+    explorationId: String
+  ): DataProvider<Any?> {
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) { explorationCheckpointDatabase ->
+
+      if (!explorationCheckpointDatabase.explorationCheckpointMap.containsKey(explorationId))
+        return@storeDataWithCustomChannelAsync Pair(
+          explorationCheckpointDatabase,
+          ExplorationCheckpointActionStatus.CHECKPOINT_NOT_FOUND
+        )
+
+      val explorationCheckpointDatabaseBuilder = explorationCheckpointDatabase.toBuilder()
+
+      explorationCheckpointDatabaseBuilder
+        .removeExplorationCheckpoint(explorationId)
+
+      Pair(
+        explorationCheckpointDatabaseBuilder.build(),
+        ExplorationCheckpointActionStatus.SUCCESS
+      )
+    }
+    return dataProviders.createInMemoryDataProviderAsync(
+      DELETE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID
+    ) {
+      return@createInMemoryDataProviderAsync getDeferredResult(
+        deferred = deferred,
+        profileId = profileId,
+        explorationId = explorationId
+      )
+    }
+  }
+
+  private suspend fun getDeferredResult(
+    deferred: Deferred<ExplorationCheckpointActionStatus>,
+    explorationId: String?,
+    profileId: ProfileId?,
+  ): AsyncResult<Any?> {
+    return when (deferred.await()) {
+      ExplorationCheckpointActionStatus.CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED ->
+        AsyncResult.success(
+          ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED
+        )
+      ExplorationCheckpointActionStatus.CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED ->
+        AsyncResult.success(
+          ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED
+        )
+      ExplorationCheckpointActionStatus.CHECKPOINT_NOT_FOUND ->
+        AsyncResult.failed(
+          ExplorationCheckpointNotFoundException(
+            "No saved checkpoint with explorationId ${explorationId!!} found for " +
+              "the profileId ${profileId!!.internalId}."
+          )
+        )
+      ExplorationCheckpointActionStatus.SUCCESS ->
+        AsyncResult.success(null)
```

Comment:
Why do we need to send null here but in rest of the case we are sending an enum value?

Reply:
null is being sent in cases where we only need to know if the operation was successful or not. (for example when we delete a checkpoint).

Reply:
Okay